packages/dds/merge-tree/src/mergeTree.ts

```diff
@@ -1302,6 +1303,74 @@ export class MergeTree {
 		return undefined;
 	}
 
+	/**
+	 * Finds the nearest reference with ReferenceType.Tile to `startPos` in the direction dictated by `posPrecedesTile`.
+	 * Uses depthFirstNodeWalk in addition to block-accelerated functionality. The search position will be included in
+	 * the nodes to walk, so searching on all positions, including the endpoints, can be considered inclusive.
+	 * Any out of bound search positions will return undefined, so in order to search the whole string, a forward
+	 * search can begin at 0, or a bacward search can begin at length-1.
+	 *
+	 * @param startPos - Position at which to start the search
+	 * @param clientId - clientId dictating the perspective to search from
+	 * @param tileLabel - Label of the tile to search for
+	 * @param backwards - Whether the desired tile comes before (true) or after (false) `startPos`
+	 */
+	public searchForTile(startPos: number, clientId: number, tileLabel: string, backwards = true) {
+		const searchInfo: IReferenceSearchInfo = {
+			mergeTree: this,
+			tilePrecedesPos: backwards,
+			tileLabel,
+		};
+
+		const { segment } = this.getContainingSegment(startPos, UniversalSequenceNumber, clientId);
+		if (segment?.parent === undefined) {
+			return undefined;
+		}
+
+		depthFirstNodeWalk(
+			segment.parent,
+			segment,
+			(seg) => {
+				if (seg.isLeaf()) {
+					if (Marker.is(seg) && refHasTileLabel(seg, searchInfo.tileLabel)) {
+						searchInfo.tile = seg;
+						return NodeAction.Exit;
+					}
+					return NodeAction.Skip;
+				}
+				const block = <IHierBlock>seg;
+				const marker = searchInfo.tilePrecedesPos
+					? <Marker>block.rightmostTiles[searchInfo.tileLabel]
+					: <Marker>block.leftmostTiles[searchInfo.tileLabel];
+				if (marker !== undefined) {
+					searchInfo.tile = marker;
+					return NodeAction.Exit;
+				}
```

Comment:
right now you aren't skipping internal nodes. I think I'd combine all of the `return`s of this function into the end with a `return searchInfo.tile !== undefined ? NodeAction.Edit : NodeAction.Skip`.